src/pgrepr/src/value/apd.rs

```diff
@@ -0,0 +1,242 @@
+            cx.div_integer(&mut w, &TO_SQL_BASER);
+            d = w;
+        }
+        d_i -= leading_zero_units;
+
+        let units = u16::try_from(UNITS_LEN - d_i).unwrap();
+        let weight = i16::try_from(units - fract_units).unwrap() - 1;
+
+        out.put_u16(units);
+        out.put_i16(weight);
+        // sign
+        out.put_u16(if is_nan {
+            0xC000
+        } else if is_neg {
+            0x4000
+        } else {
+            0
+        });
+        out.put_u16(scale);
+        if !is_nan {
+            for digit in digits[d_i..].iter() {
+                out.put_u16(*digit);
+            }
+        }
+
+        Ok(IsNull::No)
+    }
+
+    fn accepts(ty: &Type) -> bool {
+        matches!(*ty, Type::NUMERIC)
+    }
+
+    to_sql_checked!();
+}
+
+impl<'a> FromSql<'a> for Apd {
+    fn from_sql(_: &Type, mut raw: &'a [u8]) -> Result<Apd, Box<dyn Error + Sync + Send>> {
+        let units = usize::from(raw.read_u16::<NetworkEndian>()?);
+        let weight = raw.read_i16::<NetworkEndian>()?;
+        let sign = raw.read_u16::<NetworkEndian>()?;
+        let in_scale = raw.read_i16::<NetworkEndian>()?;
+        let mut digits = vec![];
+        for _ in 0..units {
+            digits.push(raw.read_u16::<NetworkEndian>()?)
+        }
+
+        // We need wider context because decoding values can require >39 digits
+        // of precision given how alignment works.
+        let mut cx = apd::cx_agg();
+        let mut d = ApdAgg::zero();
+
+        for digit in digits[..units].iter() {
+            cx.scaleb(&mut d, &FROM_SQL_SCALER);
+            let n = AdtApd::from(u32::from(*digit));
+            cx.add(&mut d, &n);
+        }
+
+        match sign {
+            0 => (),
+            0x4000 => cx.neg(&mut d),
+            0xC000 => return Ok(Apd(OrderedDecimal(AdtApd::nan()))),
+            _ => return Err("bad sign in numeric".into()),
+        }
+
+        let mut scale = (units as i16 - weight - 1) * 4;
+
+        // Adjust scales
+        if scale < 0 {
+            // Multiply by 10^scale
+            cx.scaleb(&mut d, &AdtApd::from(-i32::from(scale)));
+            scale = 0;
+        } else if scale > in_scale {
+            // Divide by 10^(difference in scale and in_scale)
+            cx.scaleb(&mut d, &AdtApd::from(-i32::from(scale - in_scale)));
+            scale = in_scale;
+        }
+
+        cx.scaleb(&mut d, &AdtApd::from(-i32::from(scale)));
+        cx.reduce(&mut d);
+
+        let mut cx = cx_datum();
+        let d_datum = cx.to_width(d);
+
+        // Reducing before taking to datum width lets us check for any status
+        // for errors.
+        if d.is_infinite() || cx.status().any() {
+            return Err(format!("Unable to take bytes to APD value; rendered {}", d).into());
+        }
+        Ok(Apd(OrderedDecimal(d_datum)))
+    }
+
+    fn accepts(ty: &Type) -> bool {
+        matches!(*ty, Type::NUMERIC)
+    }
+}
+
```

Comment:
I also came up with this unit test, but I am not sure if it even belongs to this file. If you think it is useful, you can consider pasting it in its rightful location:

```
#[test]
fn test_invalid_numbers() {
    fn inner(s: &str) {
        let mut cx = apd::cx_datum();
        let d = cx.parse(s);

        match d {
            Err(_) => {}
            _ => {
                panic!("Error expected but not thrown for {:#?}, {:#?}", s, d)
            }
        }
    }

    inner("e25");
    inner("1e");
    inner("1x25");

    inner("!1");
    inner("--1");
    inner("-1-");
    inner("1-");

    inner("1..1");
    inner("..1");
    inner("1.1.1");
}```

Reply:
Added these as `slt` tests locally; will include them in next push to this PR